core/debug/src/debug.c

```diff
@@ -230,3 +230,117 @@ void debug_fault(THaltError reason, uint32_t lr, uint32_t sp)
     debug_mpu_config();
     DEBUG_PRINT_END();
 }
+
+static void __debug_reboot(void)
+{
+    UVISOR_SVC(UVISOR_SVC_ID_DEBUG_REBOOT, "");
+}
+
+void debug_reboot(void)
+{
+    if (!g_debug_box.initialized || g_active_box != g_debug_box.box_id) {
+        HALT_ERROR(NOT_ALLOWED, "This function can only be called from the context of an initialized debug box.\n\r");
+    }
+
+    /* Reboot.
+     * This can only be called from privileged code. */
+    NVIC_SystemReset();
+}
+
+/* FIXME: Currently it is not possible to return to a regular execution flow
+ *        after the execution of the debug box handler. It is possible to
+ *        reboot, though. */
+static void debug_deprivilege_and_return(void *debug_handler, void *return_handler,
+                                         uint32_t a0, uint32_t a1, uint32_t a2, uint32_t a3)
+{
+    uint8_t src_id, dst_id;
+    uint32_t dst_sp_align;
+    uint32_t *src_sp, *dst_sp;
+
+    /* Gather information from the current state.
+     * The destination box is always the debug box. */
+    /* Note: The source stack pointer is only used to assess the stack
+     *       alignment. */
+    src_sp = svc_cx_validate_sf((uint32_t *) __get_PSP());
+    src_id = g_active_box;
+    dst_sp = g_svc_cx_curr_sp[g_debug_box.box_id];
+    dst_id = g_debug_box.box_id;
+
+    /* Check source and destination box IDs */
+    if (src_id == dst_id) {
+        HALT_ERROR(NOT_ALLOWED, "The system is already running in the context of the debug box.\n\r");
+    }
+
+    /* Create the exception stack frame. */
+    dst_sp_align = ((uint32_t) dst_sp & 0x4) ? 1 : 0;
+    dst_sp -= (SVC_CX_EXC_SF_SIZE + dst_sp_align);
+
+    /* Populate the exception stack frame:
+     *   - The scratch registers are populated with the handler arguments.
+     *   - r12 is cleared.
+     *   - lr is the thunk function (die).
+     *   - The return address is the destination function. */
+    dst_sp[0] = a0;
+    dst_sp[1] = a1;
+    dst_sp[2] = a2;
+    dst_sp[3] = a3;
+    dst_sp[4] = 0;
+    dst_sp[5] = (uint32_t) return_handler;
+    dst_sp[6] = (uint32_t) debug_handler;
+    dst_sp[7] = src_sp[7] | (dst_sp_align << 9);
+
+    /* Set the context stack pointer for the destination box. */
+    *(__uvisor_config.uvisor_box_context) = g_svc_cx_context_ptr[dst_id];
+
+    /* Switch boxes. */
+    vmpu_switch(src_id, dst_id);
+    __set_PSP((uint32_t) dst_sp);
+
+    /* Upon execution return debug_handler will be executed. Upon return from
+     * debug_handler, return_handler will be executed. */
+    return;
+}
+
+void debug_get_version(void)
+{
+    HALT_ERROR(NOT_IMPLEMENTED, "This handler is not implemented yet. Only version 0 is supported.\n\r");
```

Comment:
@Patater - agreed with @AlessandroA - debug_get_version is a placeholder now for technical reasons. It might be helpful though to explain the reasoning in a FIXME-comment